src/linear_algebra/quadratic_form.lean

```diff
@@ -0,0 +1,290 @@
+      by convert Q.polar_add_left (a • x) (a • x') (a • y) using 1; simp [polar, smul_add],
+    polar_smul_left := λ b x y,
+      by convert Q.polar_smul_left b (a • x) (a • y) using 1; simp [polar, smul_add, smul_comm],
+    polar_add_right := λ x y y',
+      by convert Q.polar_add_right (a • x) (a • y) (a • y') using 1; simp [polar, smul_add],
+    polar_smul_right := λ b x y,
+      by convert Q.polar_smul_right b (a • x) (a • y) using 1; simp [polar, smul_add, smul_comm] } ⟩
+
+@[simp] lemma smul_apply (a : R₁) (Q : quadratic_form R₁ M) (x : M) : (a • Q) x = Q (a • x) := rfl
+
+instance : mul_action R₁ (quadratic_form R₁ M) :=
+{ mul_smul := λ a b Q, ext (λ x, by simp [mul_smul, smul_comm]),
+  one_smul := λ Q, ext (λ x, by simp),
+  ..quadratic_form.has_scalar }
+
+section comp
+
+variables {N : Type v} [add_comm_group N] [module R N]
+
+/-- Compose the quadratic form with a linear function. -/
+def comp (Q : quadratic_form R N) (f : M →ₗ[R] N) :
+  quadratic_form R M :=
+{ map := λ x, Q (f x),
+  map_smul := λ a x, by simp [apply_smul],
+  polar_add_left := λ x x' y,
+    by convert Q.polar_add_left (f x) (f x') (f y) using 1; simp [polar],
+  polar_smul_left := λ a x y,
+    by convert Q.polar_smul_left a (f x) (f y) using 1; simp [polar],
+  polar_add_right := λ x y y',
+    by convert Q.polar_add_right (f x) (f y) (f y') using 1; simp [polar],
+  polar_smul_right := λ a x y,
+    by convert Q.polar_smul_right a (f x) (f y) using 1; simp [polar] }
+
+@[simp] lemma comp_apply (Q : quadratic_form R N) (f : M →ₗ[R] N) (x : M) :
+  (Q.comp f) x = Q (f x) := rfl
+
+end comp
+
+end quadratic_form
+
+/-! ### Associated bilinear forms
+
+Over a commutative ring where 2⁻¹ exists, the theory of quadratic forms is
+basically identical to that of symmetric bilinear forms. The map from quadratic
+forms to bilinear forms giving this identification is called the `associated`
+quadratic form.
+-/
+
+variables {B : bilin_form R M}
+
+namespace bilin_form
+open quadratic_form
+
+lemma polar_to_quadratic_form (x y : M) : polar (λ x, B x x) x y = B x y + B y x :=
+by simp [polar, add_left, add_right, sub_eq_add_neg _ (B y y), add_comm (B y x) _]
+
+/-- A bilinear form gives a quadratic form by applying the argument twice. -/
+def to_quadratic_form (B : bilin_form R M) : quadratic_form R M :=
+⟨ λ x, B x x,
+  λ a x, by simp [smul_left, smul_right, mul_assoc],
+  λ x x' y, by simp [polar_to_quadratic_form, add_left, add_right, add_left_comm],
+  λ a x y, by simp [polar_to_quadratic_form, smul_left, smul_right, mul_add],
+  λ x y y', by simp [polar_to_quadratic_form, add_left, add_right, add_left_comm],
+  λ a x y, by simp [polar_to_quadratic_form, smul_left, smul_right, mul_add] ⟩
+
+@[simp] lemma to_quadratic_form_apply (B : bilin_form R M) :
+(B.to_quadratic_form : M → R) = λ x, B x x :=
+rfl
+
+end bilin_form
+
+namespace quadratic_form
+open bilin_form sym_bilin_form
+
+section associated
+variables [has_inv R₁] {B₁ : bilin_form R₁ M}
```

Comment:
Do we to assume `has_inv`? Or should we ask explicitly for an inverse of `2`? Because with `has_inv` we cannot apply this to a general local ring (even though specific examples like `Z_[p]` do have a `has_inv` instance).

Reply:
I wanted to use a typeclass because it fills in the argument automatically. It also expresses the intent that there is essentially a unique `associated` form. I've changed it into a bundled `{half : R // half * 2 = 1}` argument which at least solves the second part (compared to the unbundled version which only passes `half : R` to `associated` and adds `half * 2 = 1` in a few lemmas).

A random thought. How about defining:
```lean
class invertible [monoid R] (x : R) : Type := (inv : R) (inv_mul : inv * x = 1) (mul_inv : x * inv = 1)
```
with instances along the lines of `invertible 1`, `[division_ring R] [char_zero R] [invertible x] : invertible (bit0 x)`, `[division_ring R] [char_zero R] : invertible (bit1 x)`.

Reply:
Another option would be `[fact (is_unit 2)]`, along the lines of https://gist.github.com/cipher1024/9bd785a75384a4870b331714ec86ad6f#file-zmod_redesign-lean
On the other hand, that "fact" would not record data.

Reply:
I think making sure the definitions stay computable is a good reason to take the inverse as data, so I would prefer this to `is_unit`. (I've actually tested the `discr` function by `#eval`ing it on examples.)

Reply:
Updated to use `invertible` now that it got merged